### PR TITLE
Update SMTChecker tests with z3 4.8.12

### PR DESCRIPTION
**Pull request by leonardoalt**

### CODE REVIEW

File: .circleci/osx_install_dependencies.sh [hunk 1]
```diff
@@ -48,13 +48,15 @@ then
...
   # z3
-  wget https://github.com/Z3Prover/z3/releases/download/z3-4.8.10/z3-4.8.10-x64-osx-10.15.7.zip
-  unzip z3-4.8.10-x64-osx-10.15.7.zip
-  rm -f z3-4.8.10-x64-osx-10.15.7.zip
-  cp z3-4.8.10-x64-osx-10.15.7/bin/libz3.a /usr/local/lib
-  cp z3-4.8.10-x64-osx-10.15.7/bin/z3 /usr/local/bin
-  cp z3-4.8.10-x64-osx-10.15.7/include/* /usr/local/include
-  rm -rf z3-4.8.10-x64-osx-10.15.7
+  z3="z3-4.8.12"
+  osx="osx-10.15.7"
```
Comment (cameel) [inline .circleci/osx_install_dependencies.sh]:
```suggestion
  z3_version="z3-4.8.12"
  osx_version="osx-10.15.7"
```

File: scripts/build_emscripten.sh [hunk 2]
```diff
@@ -34,7 +34,7 @@ else
     BUILD_DIR="$1"
 fi
 
-# solbuildpackpusher/solidity-buildpack-deps:emscripten-4
+# solbuildpackpusher/solidity-buildpack-deps:emscripten-6
 docker run -v "$(pwd):/root/project" -w /root/project \
-    solbuildpackpusher/solidity-buildpack-deps@sha256:434719d8104cab47712dd1f56f255994d04eb65b802c0d382790071c1a0c074b \
+	solbuildpackpusher/solidity-buildpack-deps@sha256:092da5817bc032c91a806b4f73db2a1a31e5cc4c066d94d43eedd9f365df7154 \
```
Comment (cameel) [inline scripts/build_emscripten.sh]:
```suggestion
    solbuildpackpusher/solidity-buildpack-deps@sha256:092da5817bc032c91a806b4f73db2a1a31e5cc4c066d94d43eedd9f365df7154 \
```

File: scripts/build_emscripten.sh [hunk 3]
```diff
@@ -34,7 +34,7 @@ else
     BUILD_DIR="$1"
 fi
 
-# solbuildpackpusher/solidity-buildpack-deps:emscripten-4
+# solbuildpackpusher/solidity-buildpack-deps:emscripten-6
```
Comment (cameel) [inline scripts/build_emscripten.sh:37]: Looks like we forgot to update this last time. Could this have caused some problems?

File: test/libsolidity/smtCheckerTests/functions/getters/string.sol [hunk 4]
```diff
@@ -10,6 +10,4 @@ contract C {
 // ====
 // SMTEngine: all
 // ----
-// Warning 1218: (178-224): CHC: Error trying to invoke SMT solver.
```
Comment (cameel) [inline test/libsolidity/smtCheckerTests/functions/getters/string.sol:13]: Looks like the new z3 helps in some cases.

**Review by cameel — COMMENTED** [review]
Lots of counterexamples seem to have disappeared. I that supposed to happen?

File: scripts/build_emscripten.sh (hunk 3 above)
Comment (leonardoalt) [reply scripts/build_emscripten.sh:37]: I don't think so, seems to have been just in this comment instead of the actual hash?

**Comment by leonardoalt** [comment]
@cameel disappeared in what way? I had to remove a bunch because the MacOS version gives many different results compared to the Linux ones, very annoying...

**Comment by cameel** [comment]
Like this:
```diff
- // Warning 6328: (183-202): CHC: Assertion violation happens here.\nCounterexample:\ns1 = {x: 2, t: {y: 3, a: []}, a: [], ts: []}\n\nTransaction trace:\nC.constructor()\nState: s1 = {x: 0, t: {y: 0, a: []}, a: [], ts: []}\nC.f()
+ // Warning 6328: (183-202): CHC: Assertion violation happens here.
```

**Comment by leonardoalt** [comment]
Yea I had to remove many of those because MacOS is different.

File: scripts/build_emscripten.sh (hunk 3 above)
Comment (cameel) [reply scripts/build_emscripten.sh:37]: Ah, true!

**Comment by leonardoalt** [comment]
@cameel `b_docs` actually fails here without sphinx it seems

**Comment by leonardoalt** [comment]
Which is weird because it doesn't fail for https://github.com/ethereum/solidity/pull/11672

**Comment by cameel** [comment]
> `b_docs` actually fails here without sphinx it seems

Oh. The problem is that the docs produce quite a lot of warnings right now and `docs.sh` runs Sphinx with "treat warnings as errors" so it fails. The only reason they were not showing up on `develop` was because of old Pygments and Sphinx versions. After you removed Sphinx from the image, `pip` now actually installs latest versions of packages because it cannot fall back on older ones being already present. So this effectively solves the same problem as my #11662 but without fixing the warnings.

Sorry, I did not take that into account. In that case I see 2 solutions:
1. Review and merge #11660 and #11662 before this PR.
2. Ignore it for now and put Sphinx back in the image. I'll remove it later when the warnings are already fixed. We'll be doing EVMC update soon and I think that requires rebuilding images too so I can remove it then.

**Comment by cameel** [comment]
By the way, it does not fail in #11672 because there CI does not yet use the new images (it still has the old hashes).

**Comment by leonardoalt** [comment]
@cameel This works now with the current version of https://github.com/ethereum/solidity/pull/11672

File: .circleci/config.yml [hunk 5]
```diff
@@ -9,20 +9,20 @@ version: 2.1
...
-    # solbuildpackpusher/solidity-buildpack-deps:ubuntu1604.clang.ossfuzz-9
-    default: "solbuildpackpusher/solidity-buildpack-deps@sha256:5078e1d74ab6f4329e9218c2d8c0ebe2d42817a3d4c3c62ce887100cbe9bc739"
+    # solbuildpackpusher/solidity-buildpack-deps:ubuntu1604.clang.ossfuzz-10
+    default: "solbuildpackpusher/solidity-buildpack-deps@sha256:ea15a35f6188360b425593c83e946660ab4ea4dac9b9c3bb3629e6ed57276b1d"
   emscripten-docker-image:
     type: string
-    # solbuildpackpusher/solidity-buildpack-deps:emscripten-5
-    default: "solbuildpackpusher/solidity-buildpack-deps@sha256:d28afb9624c2352ea40f157d1a321ffac77f54a21e33a8e8744f9126b780ded4"
+    # solbuildpackpusher/solidity-buildpack-deps:emscripten-6
+    default: "solbuildpackpusher/solidity-buildpack-deps@sha256:092da5817bc032c91a806b4f73db2a1a31e5cc4c066d94d43eedd9f365df7154"
```
Comment (cameel) [inline .circleci/config.yml:25]: I see these are not the most recent hashes. I guess still the ones from before the sphinx change?

**Review by cameel — APPROVED** [review]

File: .circleci/config.yml (hunk 5 above)
Comment (leonardoalt) [reply .circleci/config.yml:25]: yes, except ubuntu2004-clang which I removed `z3-static` again

**Pull request merged**